File: 232-ImplementQueueUsingStacks/232-ImplementQueueUsingStacks.java
// Last updated: 8/16/2026, 9:51:05 PM
class MyQueue {
    /* [1,2,3] given [1,2] push 3
        3 should be at the bottom of stack
    1   1
    2 ->2
        3
    stack1: 1
            2
    stack1: 3, stack2: 2
                       1
    stack1: 3, push 2, push 1 
    */
    private Stack<Integer> s1, s2;
    // s1 is the stack saves actual values
    private int front;
    // front value is saved as a const, and be modified during push/pop

    public MyQueue() {
        s1 = new Stack<>();
        s2 = new Stack<>();
    }
    
    public void push(int x) {
        if(s1.isEmpty()) front = x;
        while(!s1.isEmpty()) {
            s2.push(s1.pop());
        }
        s1.push(x);
        while(!s2.isEmpty()) {
            s1.push(s2.pop());
        }
    }
    
    public int pop() {
        int res = s1.pop();
        if(!s1.isEmpty()) front = s1.peek();
        return res;
    }
    
    public int peek() {
       return front;
    }
    
    public boolean empty() {
        return s1.isEmpty();
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */

File: 232-ImplementQueueUsingStacks/232-ImplementQueueUsingStacks.py
# Last updated: 8/16/2026, 9:51:01 PM
class MyQueue:

    def __init__(self):
        self.s1 = deque() # queue
        self.s2 = deque() # used for "push"

    def push(self, x: int) -> None:
        # s1 = 1 2 3
        # we want s1 = 1 2 3 4
        # push all items in s1 to s2, so s2 = 3 2 1, s1 = [], push 4 to s2, s2 = 4 3 2 1
        # push all items in s2 to s1, so s1 = 1 2 3 4, s2 = []
        while self.s1:
            self.s2.append(self.s1.pop())
        
        self.s2.append(x)

        while self.s2:
            self.s1.append(self.s2.pop())

    def pop(self) -> int:
        return self.s1.pop()

    def peek(self) -> int:
        return self.s1[-1]

    def empty(self) -> bool:
        return not bool(self.s1)


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()